### ISSUE TITLE
Relax IP validation in admin screens

### ISSUE THREAD

**Issue by martinbaillie** [tailscale/tailscale#74]
It would be nice if the admin console allowed for storing more than just an IP address for DNS, e.g. DNS over TLS: 1.1.1.1@853

<img width="377" alt="Screen Shot 2020-02-18 at 9 26 56 pm" src="https://user-images.githubusercontent.com/613740/74727693-a9c3af00-5295-11ea-9a40-243772f0e3f7.png">

**Comment by danderson** [comment]
Right now we implement this DNS policy by reconfiguring the local resolver on Windows, Mac, Unixen, and iOS. Unless they all support DoT/DoH natively, we can't really make this change.

Likely the fix for this is going to be running our own DNS proxy in tailscaled, and doing "fancy" lookups there instead of delegating that to the OS. So I think this is a fine idea, but there's a bunch of stuff we need to do first.

**Comment by danderson** [comment]
Aspirationally adding to Tailscale 1.8 milestone. I suspect I won't have time to get to it, but while I'm in this code is the right time to think on it at least.

**Comment by DentonGentry** [comment]
In 1.12, when 100.100.100.100 needs to forward a lookup to a well-known global resolver like CloudFlare or dns.google, it automatically use DNS-over-HTTPS so the query doesn't go over the open Internet in the clear.

**Comment by bradfitz** [comment]
@Saklad5, that Cloudflare test page has 4 checks. Which check are you talking about?

> Is there a technical reason that Tailscale's DNS layer can't be disabled entirely? 

You can. There's a checkbox. Or CLI: `tailscale up --accept-dns=false`

**Comment by Saklad5** [comment]
I am specifically referring to the "Secure DNS" checkbox. When Tailscale's VPN is enabled, that check fails with the following note:
> We detected you’re using 1.1.1.1 (a secure DNS resolver) but not over a secure connection.

When the Tailscale VPN is disabled, the check is passed. This is what I'd expect, as I have a DNS [configuration profile](https://github.com/paulmillr/encrypted-dns/blob/master/signed/cloudflare-tls.mobileconfig) active at all times to use DNS-over-TLS.

None of the options Tailscale provides, including whether DNS settings are used or enforced, seems to impact this. Bizarrely, leaving it unenforced causes iOS to use the DNS servers provided by the current network, ignoring both Tailscale _and_ my configuration profile.

**Comment by bradfitz** [comment]
> We detected you’re using 1.1.1.1 (a secure DNS resolver) but not over a secure connection.

@Saklad5, on the machine where & when you see that, please run `tailscale bugreport` right after and share the output.

We should only be doing DoH to 1.1.1.1.

In any case, this is getting a bit tangential this bug. We should probably move it to a separate one.

**Comment by karnauskas** [comment]
I'll add this https://developers.cloudflare.com/cloudflare-one/tutorials/secure-dns-network/#configure-your-router as it similar to nextdns.

**Comment by rs** [comment]
I would like to propose an implementation for this. In the nameservers section of the DNS tab, you may specify a DoH URL. The tailscale daemon would act as a DNS proxy in order to route DNS queries to the DoH endpoint. When MagicDNS is enabled, magic queries are intercepted before going to the DoH endpoint.

As the founder of NextDNS, I would love to see an additional feature: DoH URL templates. The idea would be to allow the use of some variables to construct the DoH URL. This could be used to pass some per client dynamic information to identify the client in the DNS logs like the hostname etc. (this is an important feature for many users). This could also allow advanced DNS policy routing with conditions based on those variables.

Here is a list of variables I have in mind:
* A unique client opaque ID (if tailscale has such thing)
* The client hostname
* The client tailscale IP
* The client OS name
* The client tags (comma separated?)

Those variables could be added to the URL path, query string or headers with the proper encoding depending on the placement.

**Comment by ronaldheft** [comment]
+1 to this. Not supporting custom DoH is holding me back from switching my network to Tailscale.

I am a NextDNS user, and I like the fact that my DNS is not hosted locally, is resilient to local network outages, and still supports custom filtering rules regardless of the network I'm on.

Tailscale's mesh model is wonderful, that it's not reliant on a single local server hosting my VPN. I'd like to ensure I can have that same resiliency and speed by using a hosted DoH DNS solution alongside a Tailscale network.

**Comment by bradfitz** [comment]
@rs, the Tailscale client and coordination server already support 90% of what you describe. What remains is bootstrap DNS. The DoH providers we support today all share the same property: their DoH is served by their well known UDP/TCP DNS IP. So we don't need to do the bootstrap resolution to find which IP to send the DoH query to. That's why we don't yet support e.g. OpenDNS even though we could in theory transparently upgrade it like we so for 1.1.1.1, 8.8.8.8, 9.9.9.9, etc.

Once we add the bootstrap DoH DNS resolution then the custom DoH URL support we already have should just basically work and we'd just need to permit it in the admin console.

**Comment by rs** [comment]
Glad to hear that @bradfitz. What do you think about the DoH URL template part? Is it something you would be will to support?

**Comment by bradfitz** [comment]
Yeah, that's probably fine. We'd need some concrete use cases and to decide on which variables we support.

**Comment by rs** [comment]
An example with NextDNS, in order to pass a hostname for the profile id `abcdef`:

> For "John's iPhone", you would use https://dns.nextdns.io/abcdef/John's%20iPhone as your DNS-over-HTTPS endpoint.

This hostname is then shown in your DNS logs. Here is an example with my own logs:

<img width="823" alt="178161187-6f2a0800-d1d3-4cd8-8b76-3f4ce99ca4c3" src="https://user-images.githubusercontent.com/68232/178161304-cb065bc0-b8cb-4204-b58d-9f428eda7ba1.png">


 We currently support the following information for log client identification:
* Device unique ID
* Device Name (can be derived from hostname)
* Device model (could be derived from OS)
* Device LAN IP (would be mapped to tailscale IP)

They can all be passed via the path (except for the device IP) as follow: `/<profile id>/<hostname>/<model>/<device id>`. Alternatively, they can be passed as HTTP headers (`X-Device-Name`, `X-Device-Model`, `X-Device-Id`, `X-Device-Ip`). In all cases, HPACK compression ensures they won't inflate all DNS queries.

Another use case would be to use tailscale's tags to map a group of hosts to a NextDNS profile.

Does it make sense?

**Comment by bradfitz** [comment]
Yup, makes sense.

**Comment by DentonGentry** [comment]
NextDNS support launched in Tailscale 1.32: https://tailscale.com/blog/nextdns/

**Comment by CaptainXLAB** [comment]
Can there be support custom DoH/DoT servers other than nextdns as well?

I've already setup a VPS that all my devices connect to over HTTPS on windows and over TLS for android, configured with pihole and unbound.

When tailscale is enabled, android considers it as a "VPN" and doesn't use "Private DNS" option which is what sends all DNS requests over TLS to my own VPS.

For Windows, I'm using instantsc's fork of SimpleDNScrypt which also supports custom resolvers. I already have the DNS stamp info.

I can't "force DNS" onto the entire tailnet because the VPS itself is also configured as a tailscale client, which resolves everything over localhost, and "force DNS" also forces it's DNS resolver to change, and everything on the tailnet can't resolve dns requests after that. Also, force DNS freaks out Kaspersky AV for some reason and kaspersky simply deletes tailscale and it's setup file tagging it as a trojan. I can simply download the setup again and reinstall it, then it works fine.

"Add a custom DNSCrypt resolver" option would be a really nice feature to have in tailscale's DNS menu, with just 2 fields - DNS name and DNS stamp info (sdns://)

**Comment by Saklad5** [comment]
DNSCrypt is obsolete, anyway. If you were going to add a new standard, it'd best be something like DNS-over-QUIC and Discovery of Designated Resolvers.

**Comment by DentonGentry** [comment]
> I can't "force DNS" onto the entire tailnet because the VPS itself is also configured as a tailscale client, which resolves everything over localhost, and "force DNS" also forces it's DNS resolver to change, and everything on the tailnet can't resolve dns requests after that.

`tailscale up --accept-dns=false` on the VPS might be useful.

**Comment by bradfitz** [comment]
@CaptainXLAB, what I said about bootstrap DNS is still the issue. Turns out NextDNS didn't need that.

**Comment by jlelse** [comment]
Any updates on this? I would like to try AdGuard DNS as an alternative to NextDNS (because it's already bundled in my subscription).

**Comment by DentonGentry** [comment]
I promise we will update the bug when there is something to report. The most recent development on this issue was NextDNS.

**Comment by Glenta** [comment]
Will there also be a look into [Control D](https://controld.com/)? People have preferences on who to trust their DNS queries with.

**Comment by DentonGentry** [comment]
Feel free to file a feature request for Control D specifically.

We have more ability to accommodate specific providers within tailscaled itself. For example, Google and CloudFlare DNS IP addresses are recognized and turn into DNS-over-HTTPS automatically when tailscaled is forwarding the query. 

Where we have trouble is accepting arbitrary URLs which tailscaled itself won't always be handling them. Some of the client operating systems we support don't support DoH or DoT. For example we write out 8.8.8.8 or 1.1.1.1 to the OS config even while tailscaled will turn those into DoH requests.

**Comment by LouisStAmour** [comment]
Regarding DoH or DoT support, it might not have been widespread in 2020 but it is incredibly widespread in 2023. I have no stats on hand to back it up, but I’ve been using DoH or DoT per device and it works great on Android (11 or later), iOS (14 or later via DNS profile), Windows 10 Insider or Windows 11, macOS 11 or later, and modern nix boxes.

If there is a limitation in VPN support for DoH or DoT please clearly outline what is needed and we can start passing on feature requests to OS providers. I know it seems silly, but change here won’t happen unless we make some noise to let OS vendors know what’s missing.

Off topic, but I’m shocked that Tailscale frequently prefers custom integrations that inevitably break early rather than open standards that can be updated to support more configurations. NextDNS is a great product but DoH or DoT are open to anyone. Google is a great SSO, but OAuth could be open to anyone. Tailscale itself is great for passwordless auth, but so is WebAuthn generically or Passkeys, both of which Tailscale could support if more open-minded in features to implement.)

Linking this ticket to #7946 as implementing this would fix support for ControlD as well as cloud-hosted DNS via DoH/DoT or custom ports for a bit of security through obscurity.

Bonus points if you can configure a DoH or DoT route per device or otherwise pass device metadata through as documented above for NextDNS.

**Comment by bradfitz** [comment]
@LouisStAmour, everything we already said above is still applicable, but I'll repeat: we did 8.8.8.8 and 1.1.1.1 and such first because they could be done transparently (their TCP/HTTPS DoH endpoints are the same anycast IPs as their well-known DNS port 53 IPs, without any extra/sketchy DNS lookup required). Then NextDNS was also super easy, and we like the NextDNS folk. They're nice and move quickly and changed things we wanted changed to make it even easier for us. Arbitrary/other DoH has bootstrapping problems and we haven't done it yet. Basically no operating system has reasonable APIs for us to say "Hey $OS, use this DoH URL for this DNS suffix". Which means at that point we have to take over all DNS into 100.100.100.100 and do DoH ourselves. That's fine. But it requires a bunch of other DNS plumbing work we've been slowly doing across all operating systems.

> Off topic, but I’m shocked that Tailscale frequently prefers custom integrations that inevitably break early 

I don't even know what that means.
 
> Google is a great SSO, but OAuth could be open to anyone

You must've missed the news: https://tailscale.com/kb/1240/sso-custom-oidc/ ... we support any OpenID Connect provider now.

This bug remains open. We'll share more when we have something to share. Trying to neg us about it won't help it get done faster.

**Comment by LouisStAmour** [comment]
Thanks for the quick reply and clarifications. Glad to see the new OIDC support, I did miss that. 

It sounds like the problem here is delayed primarily because at a networking layer, DoH and DoT require a DNS lookup (layer 7) in order to configure routing at the TCP layer for DNS requests (layer 3). If you ask me, using DNSSEC or a favoured DoH provider to look up the custom DoH endpoint seems to make sense. Cache it for as long as the TTL, and you’re golden. 

Workaround idea: It is more work than hardcoding an anycast IP, but if it’s a blocker, maybe you could request all custom DoH URLs use only IP addresses in their URLs in the short term and ignore DoT and that “sketchy” DNS lookup. It would still allow us to build custom URLs for third party DoH services now rather than wait for a perfect implementation and would be backward compatible to future implementations that support domain names or DoT again.

I’m not trying to bug anyone. Just trying to get to any workable solution. There are a lot of DoH resolvers: https://github.com/curl/curl/wiki/DNS-over-HTTPS#publicly-available-servers

**Comment by bradfitz** [comment]
> If you ask me, using DNSSEC or a favoured DoH provider

By policy, we never hard-code Google or Cloudflare or any other provider. Users have strong opinions about such things. The only one we might ever hard-code is our own IPs, but we're putting off running a global DNS service (even a limited bootstrap one) for as long as possible. It's not super fun.

> maybe you could request all custom DoH URLs use only IP addresses in their URLs in the short term and ignore DoT and that “sketchy” DNS lookup

DoH/DoT is TLS. That means you need a cert for the client to verify. You can't ~really (easily) get certs for raw IP addresses (normal people can't). And if you don't have an IP address, then you need to look up a hostname for the DoH URL. We hard code a bunch of well-known popular ones: https://github.com/tailscale/tailscale/blob/b1248442c3e4/net/dns/publicdns/publicdns.go#L133-L188

But if you want an arbitrary one, then somebody needs to do that bootstrap DNS lookup. We don't want to look it up server-side because it might involve any anycast DNS server along the way and give you geo-specific DNS results.

One option we've considered is adding a basic recursive resolver into tailscaled.

Another option is letting users specify a raw DoH URL along with a bootstrap provider: "pick which you like the most: Google, Cloudflare, Quan9, or Mullvad".


> 
Thanks for the quick reply and clarifications. Glad to see the new OIDC support, I did miss that.

It sounds like the problem here is delayed primarily because at a networking layer, DoH and DoT require a DNS lookup (layer 7) in order to configure routing at the TCP layer for DNS requests (layer 3). If you ask me, using DNSSEC or a favoured DoH provider to look up the custom DoH endpoint seems to make sense. Cache it for as long as the TTL, and you’re golden.

Workaround idea: It is more work than hardcoding an anycast IP, but if it’s a blocker, maybe you could request all custom DoH URLs use only IP addresses in their URLs in the short term and ignore DoT and that “sketchy” DNS lookup. It would still allow us to build custom URLs for third party DoH services now rather than wait for a perfect implementation and would be backward compatible to future implementations that support domain names or DoT again.

> I’m not trying to bug anyone. Just trying to get to any workable solution. There are a lot of DoH resolvers: https://github.com/curl/curl/wiki/DNS-over-HTTPS#publicly-available-servers

You're doing it again. Do you think we don't know that or can't find that? You're negging. Please stop. That's not how you make friends & influence people on the internet.

**Comment by LouisStAmour** [comment]
Good point re workaround. TLS for an IP sounds like something Let’s Encrypt could support but with more risk than DNS without the involvement of ASIN or something.

+1 for bootstrap resolver option, though if you don’t hardcode it seems like you still need to resolve?

I’ll try to keep any future comments concise and constructive. I’ve never heard of the term “negging” before. I’m not sure it applies here. I only posted what I did because it sounded like this wasn’t a priority given it has been open for over two years and special integration was made for some but not all of the providers.

Given the open source nature of this project, is there anything we can do in these issue threads to help, that would be constructive? Or just sit on our hands and +1? I was indeed trying to prod for action but… I’ll stop now.

**Comment by ozzieba** [comment]
Stupid question, for bootstrapping DoH/DoT, what if you allow users to specify a combination of a URL (with a hotsname, used for SNI/Host header and to verify the cert), and an IP, which they get however they choose? Obviously not everyone will know how to do this, but for people who are configuring custom DoH servers on their VPN it seems like a reasonable ask.

**Comment by Saklad5** [comment]
Just use [RFC 9461](https://www.rfc-editor.org/rfc/rfc9461.html), which was published since this issue was opened.

**Comment by ozzieba** [comment]
@Saklad5 afaict that still doesn't let you avoid downgrade attacks by adversarial mitm, does it?

**Comment by Saklad5** [comment]
It does if you use DNSSEC. Which you should.

**Comment by ozzieba** [comment]
Yeah that's one way,  and maybe should be an option, but :
Not everyone wants to set up DNSSEC, and it seems to me that just specifying an IP is a reasonable alternative
Even with DNSSEC you are essentially leaking the information that you are looking for secure DNS, which you may not always have to do with DoH (if the server IP is shared with other services)

**Comment by ozzieba** [comment]
Yeah that's one way,  and maybe should be an option, but :
Not everyone wants to set up DNSSEC, and it seems to me that just specifying an IP is a reasonable alternative
Even with DNSSEC you are essentially leaking the information that you are looking for secure DNS, which you may not always have to do with DoH (if the server IP is shared with other services)

**Comment by mnordhoff** [comment]
OpenDNS/Umbrella supposedly supports DoH on the main hostnames and IPs now.

https://umbrella.cisco.com/blog/enhancing-support-dns-encryption-with-dns-over-https

**Comment by mrbluecoat** [comment]
> @CaptainXLAB, what I said about bootstrap DNS is still the issue. Turns out NextDNS didn't need that.

Would _Safe Surfer_ be an easy DoH add? They offer IPv4/IPv6 addresses but for per-client monitoring you need DoH or DoT: https://helpdesk.safesurfer.io/en-nz/article/use-safe-surfer-via-secure-dns-on-google-chrome-uarybn/

**Comment by 0xFFA500** [comment]
Let’s Encrypt seems to support IP address certificates, at least on the staging branch for now, I assume it'll come to prod at some point

Would this make it easy for Tailscale to add support for custom DoH/DoT servers assuming they're IP address based?

I'd really like to use my own encrypted DNS service on Tailscale, I've tried pointing to a Tailscale IP but I keep getting intermittent issues doing that, I assume due to connections going to the idle state but that's just a hunch

**Comment by Saklad5** [comment]
> Would this make it easy for Tailscale to add support for custom DoH/DoT servers assuming they're IP address based?
> 
> I'd really like to use my own encrypted DNS service on Tailscale, I've tried pointing to a Tailscale IP but I keep getting intermittent issues doing that, I assume due to connections going to the idle state but that's just a hunch

You don't need a PKI certificate to use encrypted DNS, and I really don't think it'd contribute much from a security perspective. I think DANE would make more sense for those using domains, perhaps with hardcoded keys for the few people who cannot use a domain at all.

**Comment by ghost** [comment]
Any update on this Tailscale? Can you hard code Adguards Public DNS (has DoH support) address in the meantime?

**Comment by ghost** [comment]
Hi @danderson, do you have any updates on this? Right now, you support/partner with NextDNS (which seems to have been abandoned) and ControlD (which uses Anycast and, in my case, and several others reported, routes me to the opposite side of the world for the endpoint, with them not being able to do anything about it)

I use AdGuard DNS (Private, Cloud), and I'm missing DoH (or even DNS over QUIC), because there is no option in Tailscale to add those. I don't want to add the plain, unencrypted IPs there.

Is this in the roadmap somehow? Will this ever be supported?

Thanks.
